### PR TITLE
Added e2e tests + small refactoring.

### PR DESCRIPTION
**Pull request by bwplotka**
cc @4n4nd

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
Almost there, does not work fully yet.

File: README.md [hunk 1]
```diff
@@ -25,8 +25,8 @@ Commands:
   help [<command>...]
     Show help.
 
-  export [<flags>]
-    Export Observability Data into popular analytics formats.
+  export --matchers=MATCHERS --min-time=MIN-TIME --max-time=MAX-TIME --resolution=RESOLUTION --out=OUT [<flags>]
+    Export observability fata into popular analytics formats.
```
Comment (4n4nd) [inline README.md:29]: is `fata` intentional or was it meant to be `data`?

Comment (bwplotka) [reply README.md:29]: nope, typo (: thanks for spotting

**Comment by bwplotka** [comment]
Works now well, please review (:

**Comment by bwplotka** [comment]
docker not supported on those containers.... We will need to solve it somehow, maybe moving to Circle CLI ): Let's do this in separate PRs

File: cmd/obslytics/export.go [hunk 2]
```diff
@@ -21,55 +26,42 @@ import (
...
-	cmd := app.Command("export", "Export Observability Data into popular analytics formats.")
+	cmd := app.Command("export", "Export observability data into popular analytics formats.")
 	inputFlag := extflag.RegisterPathOrContent(cmd, "input-config", "YAML for input configuration.", true)
 	// TODO(inecas): add support for output configuration.
 	outputFlag := extflag.RegisterPathOrContent(cmd, "output-config", "YAML for output configuration.", false)
 
-	seriesParams := input.SeriesParams{}
-	cmd.Flag("match", "Name of the metric to export").Required().StringVar(&seriesParams.Matcher)
+	// TODO(bwplotka): Describe more how the format looks like.
+	matchersStr := cmd.Flag("matchers", "Metric matcher for metrics to export (e.g up{a=\"1\"}").Required().String()
```
Comment (iNecas) [inline cmd/obslytics/export.go:35]: What's the use-case for multiple matchers?

Comment (bwplotka) [reply cmd/obslytics/export.go:35]: Essentially OR to have multiple metrics in parquet file. Also I don't get why we treat __name__ label (metric name) specially in parquet. Don't we have to add that to file as well?

Comment (iNecas) [reply cmd/obslytics/export.go:35]: I'm not sure I can think of a good use-case for `OR` on multiple metrics. The single parquet file represents a single table with specific schema: for multiple metrics, I would rather go for two separate exports.

> Don't we have to add that to file as well?
I don't thing I understood question properly, could you rephrase?

**Comment by iNecas** [comment]
Changes make sense to me (except some clarification needed on the multiple matchers).

When trying to run the e2e, I was getting `cannot run command "docker network create" in rootless mode`. @bwplotka is there a documentation on how it's expected to run these tests?

Also, do you know what happened to our CI?

**Comment by iNecas** [comment]
Btw. tested locally and I can confirm things work

File: cmd/obslytics/export.go (hunk 2 above)
Comment (iNecas) [reply cmd/obslytics/export.go:35]: Orthogonal to those questions, `--match` would be more consistent with other tools in the ecosystem (https://manpages.debian.org/testing/prometheus/promtool.1.en.html)

File: pkg/input/input.go [hunk 3]
```diff
@@ -18,9 +17,9 @@ type InputConfig struct {
 
 // SeriesParams determines what data should be loaded from the input.
 type SeriesParams struct {
-	Matcher string
-	MinTime thanosmodel.TimeOrDurationValue
-	MaxTime thanosmodel.TimeOrDurationValue
+	Matchers []*labels.Matcher
```
Comment (4n4nd) [inline pkg/input/input.go:20]: @bwplotka @iNecas Do we know of a way to translate this `labels.Matcher` ([here](https://pkg.go.dev/github.com/prometheus/prometheus/pkg/labels@v1.8.2-0.20200629082805-315564210816?tab=doc#Matcher)) type to `prompb.LabelMatcher` ([here](https://pkg.go.dev/github.com/prometheus/prometheus/prompb@v1.8.2-0.20200629082805-315564210816?tab=doc#LabelMatcher))? or should I try to individually convert it?

Comment (4n4nd) [reply pkg/input/input.go:20]: nvm I found the function I need in storepb (https://github.com/thanos-io/thanos/blob/v0.15.0/pkg/store/storepb/custom.go#L450)

File: cmd/obslytics/export.go (hunk 2 above)
Comment (4n4nd) [reply cmd/obslytics/export.go:35]: multiple matchers to form a single query makes sense to me, but I don't get how you can do multiple metrics in a single query. You could do `{__name__="a|b"}` but that doesn't make sense in terms of other labels. 
I think we should make multiple corresponding queries for multiple metrics, this might also be a good way to keep query sizes small.

File: pkg/input/input.go (hunk 3 above)
Comment (bwplotka) [reply pkg/input/input.go:20]: Yes (:

File: cmd/obslytics/export.go (hunk 2 above)
Comment (bwplotka) [reply cmd/obslytics/export.go:35]: >> Don't we have to add that to file as well?

> I don't thing I understood question properly, could you rephrase?

To me if I read this correctly: https://parquet.apache.org/documentation/latest/#file-format, this file expects columns. Metric name is just another column `__name__`. Given also that this recommendation: https://www.dremio.com/tuning-parquet/#:~:text=The%20official%20Parquet%20documentation%20recommends,SET%20command%20on%20the%20store. I expect those file to be up to 1GB, which means even thousands of metrics in single parquet file.

Also if we skip metric name in parquet file how do we even know what metric name this parquet file corresponds to?

I feel it's expected from this to actually add the label __name__. It's then up to obslytics user to choose if they want one or more metrics per file (: WDYT?

>  Orthogonal to those questions, --match would be more consistent with other tools in the ecosystem (https://manpages.debian.org/testing/prometheus/promtool.1.en.html)

Agree, fair point, will revert.

Comment (bwplotka) [reply cmd/obslytics/export.go:35]: > I think we should make multiple corresponding queries for multiple metrics, this might also be a good way to keep query sizes small.

Again, it's up to user what it selects. Thanos/Prometheus can handle requests spanning across hundreds of thousands of series

**Comment by bwplotka** [comment]
> When trying to run the e2e, I was getting cannot run command "docker network create" in rootless mode. @bwplotka is there a documentation on how it's expected to run these tests?

Looks like some limitation of the docker in your setup. Definitely new problem, let me know if you can figure this out. If not we can put some issue on Cortex lib (: 

In terms of CI I am actively fighting with it.

File: cmd/obslytics/export.go (hunk 2 above)
Comment (bwplotka) [reply cmd/obslytics/export.go:35]: Also this adding metric name as separate column etc - let's aim for those (if we agree on this) in separate PR, this is already big (:

**Comment by bwplotka** [comment]
CI green (: 

Please PTAL tomorrow. Thanks for review so far :+1:

File: cmd/obslytics/export.go (hunk 2 above)
Comment (iNecas) [reply cmd/obslytics/export.go:35]: I understand your question now: yes, if we allow multiple metrics in a single parquet file, the name would be required. But at least from how I see parquet files being used, it really don't see a case where I would want one parquet for multiple metrics, unless very very similar labels and diffing mainly in the metric name. Definitely not the main usages I would expect.

I'm ok though on providing the functionality and see how people are using it.

**Review by iNecas — APPROVED** [review]

**Review by iNecas — CHANGES_REQUESTED** [review]
Waiting for changing `--matches` to `--match`, we can keep the multiple args if you want.

File: cmd/obslytics/export.go (hunk 2 above)
Comment (bwplotka) [reply cmd/obslytics/export.go:35]: Cool, we can always iterate (:

**Comment by bwplotka** [comment]
Good point forgot about it.

File: cmd/obslytics/export_e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	// TODO(bwplotka): Allow clients to specify image directly via function args.
+	testutil.Ok(t, os.Setenv("THANOS_IMAGE", "quay.io/thanos/thanos:v0.15.0"))
+	prom, sidecar, err := e2ethanos.NewPrometheusWithSidecar(s.SharedDir(), s.NetworkName(), "1", defaultPromConfig("test", 0, "", ""), e2ethanos.DefaultPrometheusImage())
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(prom, sidecar))
+
+	testutil.Ok(t, prom.WaitSumMetricsWithOptions(e2e.Greater(512), []string{"prometheus_tsdb_head_samples_appended_total"}, e2e.WaitMissingMetrics))
+
+	logger := log.NewLogfmtLogger(os.Stderr)
+	t.Run("export up metric to parquet file", func(t *testing.T) {
```
Comment (4n4nd) [inline cmd/obslytics/export_e2e_test.go:89]: @bwplotka Can you add `using StoreAPI` in this run string? Just to make it clear which input method we are using

**Comment by bwplotka** [comment]
Fixed cc @iNecas

**Review by iNecas — APPROVED** [review]

**Comment by iNecas** [comment]
Getting this in, we can do more changes in further PRs

File: cmd/obslytics/export_e2e_test.go (hunk 4 above)
Comment (bwplotka) [reply cmd/obslytics/export_e2e_test.go:89]: Makes sense to add that once we have other inputs (:
